File: 0146-lru-cache/0146-lru-cache.cpp
class Node {
public:
    int key, val;
    Node *left, *right;
    Node(int _key, int _val): key(_key), val(_val), left(NULL), right(NULL) {}
};

class LRUCache {
private:
    int capacity, size;
    Node *head, *tail;
    unordered_map<int, Node*> memo;

    void insertNode(Node* node) {
        if (head == NULL) head = tail = node;
        else {
            tail->right = node;
            node->left = tail;
            tail = node;
        }
    }

    void deleteNode(Node* node) {
        if (node->left != NULL) node->left->right = node->right;
        else head = node->right;
        if (node->right != NULL) node->right->left = node->left;
        else tail = node->left;
        node->right = NULL;
        node->left = NULL;
    }
public:
    LRUCache(int _capacity): capacity(_capacity), size(0), head(NULL), tail(NULL) {}

    int get(int key) {
        if (memo.find(key) == memo.end()) return -1;
        Node* node = memo[key];
        deleteNode(node);
        insertNode(node);
        return node->val;
    }

    void put(int key, int value) {
        Node *node = new Node(key, value);
        if (memo.find(key) != memo.end()) {
            Node* existing = memo[key];
            deleteNode(existing);
            insertNode(node);
            memo[key] = node;
        } else {
            if (size == capacity) {
                auto it = memo.find(head->key);
                if (it != memo.end()) memo.erase(it);
                deleteNode(head);
                memo[key] = node;
                insertNode(node);
            } else {
                ++size;
                memo[key] = node;
                insertNode(node);
            }
        }
    }
};

File: 0146-lru-cache/0146-lru-cache.java
class Node {
    Node prev, next;
    int key, val;

    Node(int key, int val) {
        this.key = key;
        this.val = val;
        this.prev = this.next = null;
    }
}

class LRUCache {
    Node head, tail;
    int capacity, size;
    HashMap<Integer, Node> map;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);
        head.next = tail;
        tail.prev = head;
        this.map = new HashMap<>();
    }

    public int get(int key) {
        if (!map.containsKey(key)) return -1;
        Node node = map.get(key);
        int value = node.val;
        Node prev = node.prev, next = node.next;
        prev.next = next;
        next.prev = prev;
        node.prev = head;
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
        map.put(key, node);
        return value;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            Node node = map.get(key);
            node.val = value;
            map.put(key, node);
            Node prev = node.prev, next = node.next;
            prev.next = next;
            next.prev = prev;
            node.prev = head;
            node.next = head.next;
            head.next.prev = node;
            head.next = node;
        } else {
            Node node = new Node(key, value);
            map.put(key, node);
            node.prev = head;
            node.next = head.next;
            head.next.prev = node;
            head.next = node;
            ++size;
        }
        if (size > capacity) {
            int evictKey = tail.prev.key;
            Node evictNode = tail.prev;
            evictNode.prev.next = tail;
            tail.prev = evictNode.prev;
            map.remove(evictKey);
            --size;
        }
    }
}